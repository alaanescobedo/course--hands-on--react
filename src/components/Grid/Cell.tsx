import styled from "@emotion/styled";
import { FC } from "react";
import { Cell as CellType, CellState } from "@/helpers/Field";

export interface CellProps {
  children: CellType;
}

export const Cell: FC<CellProps> = ({ children }) => {
  switch (children) {
    case CellState.empty:
      return <EmptyFrame />;
    case CellState.hidden:
      return <ClosedFrame />;
    default:
      return <ClosedFrame />;
  }
};

const ClosedFrame = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
  user-select: none;
  cursor: pointer;
  width: 1.8vw;
  height: 1.8vw;
  background-color: #d1d1d1;
  border: 0.6vh solid transparent;
  border-color: #fff #9e9e9e #9e9e9e #fff;
  &:hover {
    filter: brightness(1.1);
  }
`;

const EmptyFrame = styled(ClosedFrame)`
  border-color: #dddddd;
  cursor: default;
  &:hover {
    filter: brightness(1);
  }
`;
