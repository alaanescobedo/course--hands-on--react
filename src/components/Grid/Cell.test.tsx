import { render, screen, fireEvent, createEvent } from "@testing-library/react";

import { Coords, CellState } from "@/helpers/Field";
import { Cell, isActiveCell } from './Cell';

describe("Cell Component", () => {
  const coords: Coords = [1, 1];
  for (let cell = CellState.empty; cell <= CellState.weakFlag; cell++) {
    test("check prevent default contextMenu for every type of cell", () => {
      const props = {
        coords,
        onClick: jest.fn(),
        onContextMenu: jest.fn(),
      };

      render(<Cell {...props}>{cell}</Cell>);

      const cellComponent = screen.getByTestId(`${cell}_${coords}`);

      const contextMenuEvent = createEvent.contextMenu(cellComponent);

      fireEvent(cellComponent, contextMenuEvent);

      expect(contextMenuEvent.defaultPrevented).toBe(true);
    });
    test("onClick and onContextMenu hanler should be called only for active cells", () => {
      const props = {
        coords,
        onClick: jest.fn(),
        onContextMenu: jest.fn(),
      };

      render(<Cell {...props}>{cell}</Cell>);

      const cellComponent = screen.getByTestId(`${cell}_${coords}`);

      fireEvent.click(cellComponent);
      fireEvent.contextMenu(cellComponent);

      if (isActiveCell(cell)) {
        expect(props.onClick).toHaveBeenCalled();
        expect(props.onContextMenu).toHaveBeenCalled();
      } else {
        expect(props.onClick).not.toHaveBeenCalled();
        expect(props.onContextMenu).not.toHaveBeenCalled();
      }
    });
  }
});
