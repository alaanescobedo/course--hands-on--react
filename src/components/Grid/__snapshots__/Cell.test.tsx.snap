// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cell Component Cell renders correctly 1`] = `
<DocumentFragment>
  <div
    class="css-1ne0sj1"
    data-testid="0_1,1"
  />
</DocumentFragment>
`;

exports[`Cell Component Cell renders correctly 2`] = `
<DocumentFragment>
  <div
    class="css-1izl0f7"
    data-testid="1_1,1"
  >
    1
  </div>
</DocumentFragment>
`;

exports[`Cell Component Cell renders correctly 3`] = `
<DocumentFragment>
  <div
    class="css-b0ro4s"
    data-testid="2_1,1"
  >
    2
  </div>
</DocumentFragment>
`;

exports[`Cell Component Cell renders correctly 4`] = `
<DocumentFragment>
  <div
    class="css-16gkrtm"
    data-testid="3_1,1"
  >
    3
  </div>
</DocumentFragment>
`;

exports[`Cell Component Cell renders correctly 5`] = `
<DocumentFragment>
  <div
    class="css-1g8ta8o"
    data-testid="4_1,1"
  >
    4
  </div>
</DocumentFragment>
`;

exports[`Cell Component Cell renders correctly 6`] = `
<DocumentFragment>
  <div
    class="css-iv9yjv"
    data-testid="5_1,1"
  >
    5
  </div>
</DocumentFragment>
`;

exports[`Cell Component Cell renders correctly 7`] = `
<DocumentFragment>
  <div
    class="css-1bty0a"
    data-testid="6_1,1"
  >
    6
  </div>
</DocumentFragment>
`;

exports[`Cell Component Cell renders correctly 8`] = `
<DocumentFragment>
  <div
    class="css-1n2xt1p"
    data-testid="7_1,1"
  >
    7
  </div>
</DocumentFragment>
`;

exports[`Cell Component Cell renders correctly 9`] = `
<DocumentFragment>
  <div
    class="css-1sztfhz"
    data-testid="8_1,1"
  >
    8
  </div>
</DocumentFragment>
`;

exports[`Cell Component Cell renders correctly 10`] = `
<DocumentFragment>
  <div
    class="css-c52ynp"
    data-testid="9_1,1"
  >
    <div
      class="css-3tubcw"
    />
  </div>
</DocumentFragment>
`;

exports[`Cell Component Cell renders correctly 11`] = `
<DocumentFragment>
  <div
    class="css-17s3mwu"
    data-testid="10_1,1"
  />
</DocumentFragment>
`;

exports[`Cell Component Cell renders correctly 12`] = `
<DocumentFragment>
  <div
    class="css-17s3mwu"
    data-testid="11_1,1"
  >
    <div
      class="css-1w1u3hh"
    />
  </div>
</DocumentFragment>
`;

exports[`Cell Component Cell renders correctly 13`] = `
<DocumentFragment>
  <div
    class="css-17s3mwu"
    data-testid="12_1,1"
  >
    <div
      class="css-1wbzzpj"
    />
  </div>
</DocumentFragment>
`;
